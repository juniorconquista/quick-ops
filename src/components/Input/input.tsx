import React, { memo } from 'react';
import { P } from 'quarks-ui';
import styled from 'styled-components';
import { variant, VariantArgs, color, space } from 'styled-system';

import { InputProps, InputStyledProps } from './input.types';

/* TODO This component was created in the local repository as it has not yet been 
implemented in lib Quarks UI, however in the future it can be changed to be used directly from lib. */

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            width: '100%',
            label: {
                m: '12px 0px',
                color: 'text.contrast',
                fontWeight: 'bold',
                display: 'block',
            },
            input: {
                height: '36px',
                borderRadius: '6px',
                color: '#6b7480 !important',
                border: '1px solid',
                borderColor: '#dfe2e6',
                p: '12px 6px',
                display: 'flex',
                alignItems: 'center',
                bg: 'white.default',
                cursor: 'pointer',
                width: '100%',
                transition: 'boxShadow .25s ease',
                fontFamily: 'Nunito Sans, sans-serif',
                '&:hover, &:focus': {
                    boxShadow: '0 0 2px 3px #6b748090',
                    outline: 'none',
                },
            },
        },
        get success() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                label: {
                    ...base.label,
                    color: 'success.default',
                },
                input: {
                    ...base.input,
                    borderColor: 'success.default',
                    color: 'success.default',
                    '&:hover, &:focus': {
                        boxShadow: (props: {
                            colors: { success: { default: string } };
                        }) => `0 0 2px 3px ${props.colors.success.default}90`,
                        outline: 'none',
                    },
                },
            };
        },
        get danger() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                label: {
                    ...base.label,
                    color: 'danger.default',
                },
                input: {
                    ...base.input,
                    borderColor: 'danger.default',
                    color: 'danger.default',
                    '&:hover, &:focus': {
                        boxShadow: (props: {
                            colors: { danger: { default: string } };
                        }) => `0 0 2px 3px ${props.colors.danger.default}90`,
                        outline: 'none',
                    },
                },
            };
        },
    },
} as VariantArgs);

const Style: React.FC<InputStyledProps> = styled.div<InputStyledProps>`
    ${variants}
    ${color}
    ${space}
`;

const Input: React.FC<InputProps<string>> = ({
    styling = 'base',
    type = 'text',
    icon,
    placeholder,
    label,
    name,
    value,
    onChange,
    error,
    ...props
}) => (
    <Style styling={styling} {...props}>
        {label && <label htmlFor={name}>{label}</label>}
        <input
            placeholder={placeholder}
            name={name}
            id={name}
            value={value}
            type={type}
            onChange={onChange}
        />
        {error && (
            <P styling="base" mt="12px" color="danger.default">
                {error}
            </P>
        )}
    </Style>
);

export default memo(Input);
