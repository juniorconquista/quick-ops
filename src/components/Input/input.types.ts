import { ChangeEvent, ReactNode } from 'react';
import { ColorProps, SpaceProps } from 'styled-system';

export type InputStyledProps = ColorProps &
    SpaceProps & {
        /** Each styling variant has unique properties that are inherited through the type informed */
        styling?: 'base' | 'success' | 'danger';
    };

export interface InputProps<Value> extends InputStyledProps {
    name: string;
    icon?: ReactNode;
    placeholder?: string;
    label?: string;
    value?: Value;
    error?: string | boolean;
    type?: string;
    onChange?: (e: ChangeEvent<any>) => void;
    onBlur?: (e: ChangeEvent<any>) => void;
}
