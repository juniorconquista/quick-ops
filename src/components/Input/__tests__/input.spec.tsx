import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render } from '@testing-library/react';
import ThemeProvider from '../../../utils/ThemeProviderTest';

import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';

import Input from '../input';

describe('Input component', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });
    test('basic snapshot', () => {
        const container = render(
            <ThemeProvider>
                <Input styling="base" name="test" />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should inherit the styling properties of the success variant', () => {
        const input = renderer
            .create(
                <ThemeProvider>
                    <Input styling="success" name="test" />
                </ThemeProvider>,
            )
            .toJSON();
        expect(input).toHaveStyleRule('color', '#51c1c3', {
            modifier: 'label',
        });
        expect(input).toHaveStyleRule('color', '#51c1c3', {
            modifier: 'input',
        });
        expect(input).toHaveStyleRule('border-color', '#51c1c3', {
            modifier: 'input',
        });
    });

    it('should inherit the styling properties of the danger variant', () => {
        const input = renderer
            .create(
                <ThemeProvider>
                    <Input styling="danger" name="test" />
                </ThemeProvider>,
            )
            .toJSON();
        expect(input).toHaveStyleRule('color', '#ff4138', {
            modifier: 'label',
        });
        expect(input).toHaveStyleRule('color', '#ff4138', {
            modifier: 'input',
        });
        expect(input).toHaveStyleRule('border-color', '#ff4138', {
            modifier: 'input',
        });
    });
});
