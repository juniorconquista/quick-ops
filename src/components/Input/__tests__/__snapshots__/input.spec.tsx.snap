// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component basic snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
}

.c0 label {
  margin: 12px 0px;
  color: #FFFFFF;
  font-weight: bold;
  display: block;
}

.c0 input {
  height: 36px;
  border-radius: 6px;
  color: #6b7480 !important;
  border: 1px solid;
  border-color: #dfe2e6;
  padding: 12px 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #25262a;
  cursor: pointer;
  width: 100%;
  -webkit-transition: boxShadow .25s ease;
  transition: boxShadow .25s ease;
  font-family: Nunito Sans,sans-serif;
}

.c0 input:hover,
.c0 input:focus {
  box-shadow: 0 0 2px 3px #6b748090;
  outline: none;
}

<div
    class="c0"
  >
    <input
      id="test"
      name="test"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
